source/client-side-encryption/client-side-encryption.rst

```diff
@@ -770,73 +942,299 @@ behind-the-scenes operations required for encryption or decryption.
+   `libmongocrypt_handle` that is already using csfle.
+
+Drivers SHOULD document this limitation for users along with the documentation
+on the ``csfle*`` options in `cse.extraoptions`. Specifically, care should
+be taken to always specify the same set of search and override paths for the
+lifetime of a `libmongocrypt_handle` that has csfle open.
+
+Once all open handles to a `csfle` library are closed, it is possible to load a
+different `csfle` library than was previously loaded. The restriction only
+applies to simultaneous open handles within a single process.
+
+
+.. _cse.managing-mongocryptd:
+
 Managing mongocryptd
 ====================
-If a MongoClient is configured for Client Side Encryption 
-(eg. bypassAutoEncryption=false), then by default 
-(unless mongocryptdBypassSpawn=true), mongocryptd MUST be
-spawned by the driver. Spawning MUST include the command line argument
---idleShutdownTimeoutSecs. If the user does not supply one through
-extraOptions.mongocryptdSpawnArgs (which may be either in the form
-"--idleShutdownTimeoutSecs=60" or as two consecutive arguments
-["--idleShutdownTimeoutSecs", 60], then the driver MUST append
---idleShutdownTimeoutSecs=60 to the arguments. This tells mongocryptd
-to automatically terminate after 60 seconds of non-use. The stdout
-and stderr of the spawned process MUST not be exposed in the driver (e.g.
-redirect to /dev/null). Users can pass the argument --logpath to
-extraOptions.mongocryptdSpawnArgs if they need to inspect mongocryptd
-logs.
-
-Upon construction, the MongoClient MUST create a MongoClient to
-mongocryptd configured with serverSelectionTimeoutMS=10000.
-
-If spawning is necessary, the driver MUST spawn mongocryptd whenever
-server selection on the MongoClient to mongocryptd fails. If the
-MongoClient fails to connect after spawning, the server selection error
-is propagated to the user.
-
-Single-threaded drivers MUST connect with `serverSelectionTryOnce=false <../server-selection/server-selection.rst#serverselectiontryonce>`_
-, connectTimeoutMS=10000, and MUST bypass `cooldownMS <../server-discovery-and-monitoring/server-discovery-and-monitoring.rst#cooldownms>`__ when connecting to mongocryptd. See `Why are serverSelectionTryOnce and cooldownMS disabled for single-threaded drivers connecting to mongocryptd?`_.
-
-If the ClientEncryption is configured with mongocryptdBypassSpawn=true,
-then the driver is not responsible for spawning mongocryptd. If server
-selection ever fails when connecting to mongocryptd, the server
-selection error is propagated to the user.
+
+If the following conditions are met:
+
+- The user's ``MongoClient`` is configured for client-side encryption (i.e.
+  ``bypassAutoEncryption != false``)
+- **AND** the user has not disabled ``mongocryptd`` spawning (i.e. with
+  ``mongocryptdBypassSpawn=true``)
+- **AND** the `cse.csfle` library is unavailable *OR* is disabled.
+
+**then** ``mongocryptd`` MUST be spawned by the driver.
```

Comment:
Should there be a way to tell the driver “use a CSFLE shared library from a search path or fail”, i.e. to not fall back to mongocryptd? Right now, with `{ extraOptions: { mongocryptdBypassSpawn: true, csfleSearchPathsPrefer: ['/path1', '/path2'] } }`, you could tell the driver to search `/path1` and `/path2` for the shared library and not spawn mongocryptd, but you wouldn’t prevent it from attempting to connect to mongocryptd.

Reply:
Good point. I think this is a useful toggle to offer, although it will now require a few tweaks to the `libmongocrypt` library since I had not foreseen this.

Reply:
Not sure what I was thinking earlier. This can be handled entirely within the driver since `libmongocrypt` isn't the one responsible for the spawning.